File: Next Permutation.cpp
class Solution {
public:
    
    void nextPermutation(vector<int> &num) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<int>::iterator i, j, k;
        i = num.end() - 2;
        j = k = num.end() - 1;
        while (i >= num.begin() && *i >= *j) {
            i--;
            j--;
        }
        
        if (i < num.begin()) {
            reverse(num.begin(), num.end());
            return;
        }
        
        while (*i >= *k) {
            k--;
        }
        
        swap(*i, *k);
        reverse(j, num.end());
    }
};

File: Next Permutation.cs
public class Solution {
    public void NextPermutation(int[] nums) {
        if (nums == null || nums.Length <= 1) {
            return;
        }
        
        var i = nums.Length - 2;
        var j = nums.Length - 1;
        while (i >= 0 && nums[i] >= nums[j]) {
            i--;
            j--;
        }
        
        if (i < 0) {
            Array.Reverse(nums);
            return;
        }
        
        var k = nums.Length - 1;
        while (nums[i] >= nums[k]) {
            k--;
        }
        var temp = nums[i];
        nums[i] = nums[k];
        nums[k] = temp;
        Array.Reverse(nums, j, nums.Length - j);
    }
}
